### PR TITLE
Test data bleach

### PR DESCRIPTION
**Pull request by jessebrennan**
This resolves https://ucsc-cgl.atlassian.net/browse/TLP-747

### CODE REVIEW

File: tests/test_data_bleach.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+class TestDataBleach(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.test_path = Path(__file__).parents[0]
+        cls.test_file = cls.test_path / 'test_data/bleach_test_data.json'
+        cls.out_file = cls.test_path / f'{str(uuid.uuid4())}.tmp.json'
+
+    def setUp(self):
+        with self.assertRaises(FileNotFoundError):
```
Comment (mikebaumann) [inline tests/test_data_bleach.py]: What do you think of having `setUp` remove the file (if it is there) as is done in `tearDown`?
That way if the file is there, it will be removed and the test can proceed rather than stopping.
Given the name of the `out_file` contains a unique UUID, I don't think there is concern with interfering with an existing test execution.

File: tests/test_data_bleach.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                        # These are the values that the sanitizer may set to
+                        assert metadata[field][sub_field] in ['', "--------", False, 0, 0.0, []]
+
+    def _validate_output(self):
+        # check the output against the agreed upon schema!
+        assert os.path.isfile(str(self.out_file))
+        output_json = self._get_output_json()
+        for bundle in output_json:
+            metadata = bundle['data_bundle']['user_metadata']
+            # TODO: make some asserts here
```
Comment (mikebaumann) [inline tests/test_data_bleach.py]: Is this TODO still current/applicable?

File: tests/test_data_bleach.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    def _properly_sanitized(self, metadata):
+        for field in metadata:
+            if type(metadata[field]) != dict:
+                # the field should be cleared (sanitized)
+                assert not metadata[field]
+            else:
+                for sub_field in metadata[field]:
+                    if self._field_is_public(field, sub_field):
+                        # public fields have data
+                        assert metadata[field][sub_field]
```
Comment (mikebaumann) [inline tests/test_data_bleach.py]: What about the case where field is public but the value is undefined (`null`, `None`)?
It seems that is a common case that we probably should include in this unit test, and if we had test input data which contained public fields that were `null`/`None` this line in `_properly_sanitized` would fail inappropriately.

File: tests/test_data_bleach.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                    elif type(metadata[field][sub_field]) != dict:
+                        # TODO: currently, if a dict is a protected field its subfields are still
```
Comment (mikebaumann) [inline tests/test_data_bleach.py]: I think if a dict is a protected field then sanitizing its subfields IS the desired behavior (as is currently the behavior, from what I understand from your comment).

File: tests/test_data_bleach.py (hunk 1 above)
Comment (jessebrennan) [reply tests/test_data_bleach.py]: I think that's a fine implementation, but the reason that I made it like this was so that if the file isn't getting deleted, then the tests will fail so that you know something is wrong. To me this is better than adding the redundancy of trying to delete the file in multiple places.

Either way it won't affect how things work all that much.

File: tests/test_data_bleach.py (hunk 4 above)
Comment (jessebrennan) [reply tests/test_data_bleach.py]: Sorry, that's a typo. I meant to say if the dict is a **PUBLIC** field.

File: tests/test_data_bleach.py (hunk 3 above)
Comment (jessebrennan) [reply tests/test_data_bleach.py]: True, I didn't think about that. Are there any reasonable assertions we can make about these fields? I'm leaning towards no...

File: tests/test_data_bleach.py (hunk 1 above)
Comment (mikebaumann) [reply tests/test_data_bleach.py]: Sounds good.

File: tests/test_data_bleach.py (hunk 3 above)
Comment (mikebaumann) [reply tests/test_data_bleach.py]: There are reasonable assertions you can/should make if you know your test input data, which you do. You could have one test input in which the values are defined, and another test input in which the values are undefined.

File: tests/test_data_bleach.py (hunk 4 above)
Comment (mikebaumann) [reply tests/test_data_bleach.py]: Okay, the fields to be retained are specified by fully qualified names of the (non-dict) fields. Clearing all other fields seems appropriate.
